connect/src/com/telenor/mobileconnect/MobileConnectSdkProfile.java

```diff
@@ -0,0 +1,317 @@
+    }
+
+    @Override
+    public String getClientSecret() {
+        return operatorDiscoveryResult.getClientSecret();
+    }
+
+    @Override
+    public String getRedirectUri() {
+        return operatorDiscoveryConfig.getOperatorDiscoveryRedirectUri();
+    }
+
+    @Override
+    public String getExpectedIssuer(String actualIssuer) {
+        // discrepancy between .well-known configuration and the actual issuer returned
+        if (operatorDiscoveryResult.getBasePath().contains("telenordigital.com")) {
+            return actualIssuer;
+        }
+        if (getWellKnownConfig() != null) {
+            return getWellKnownConfig().getIssuer();
+        }
+        return actualIssuer;
+    }
+
+    @Override
+    public List<String> getExpectedAudiences(List<String> actualAudiences) {
+        return new ArrayList<>(actualAudiences);
+    }
+
+    @Override
+    public Uri getAuthorizeUri(Map<String, String> parameters, List<String> locales) {
+        Uri.Builder builder = ConnectUrlHelper.getAuthorizeUriStem(
+                parameters,
+                getClientId(),
+                getRedirectUri(),
+                locales,
+                getApiUrl())
+                .buildUpon();
+        for (String pathSeg : getApiUrl().pathSegments()) {
+            builder.appendPath(pathSeg);
+        }
+        return builder.build();
+    }
+
+    @Override
+    public DoNext onStartAuthorization(Map<String, String> parameters) {
+        OperatorDiscoveryAPI.OperatorDiscoveryResult odResult = null;
+
+        final String msisdn = readPhoneNumber();
+        if (msisdn != null) {
+            odResult = getOperatorDiscoveryResult(msisdn);
+        }
+
+        if (odResult != null) {
+            parameters.put("login_hint", String.format("ENCR_MSISDN:%s", odResult.getSubscriberId()));
+        }
+
+        if (isInitialized || initialize(odResult)) {
+            return DoNext.proceed;
+        }
+
+        return DoNext.cancel;
+    }
+
+    private String readPhoneNumber() {
+        TelephonyManager phMgr = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+        try {
+            String msisdn = phMgr.getLine1Number();
+            return (TextUtils.isEmpty(msisdn) ? null : msisdn);
+        } catch (RuntimeException ignored) {
+        }
+        return null;
+    }
+
+    private OperatorDiscoveryAPI.OperatorDiscoveryResult getOperatorDiscoveryResult() {
+        TelephonyManager phMgr = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+        String networkOperator = phMgr.getNetworkOperator();
+        if (!TextUtils.isEmpty(networkOperator)) {
+            final String mcc = networkOperator.substring(0, 3);
+            final String mnc = networkOperator.substring(3);
+            Future<OperatorDiscoveryAPI.OperatorDiscoveryResult> operatorDiscoveryResultFuture =
+                    sExecutor.submit(new Callable<OperatorDiscoveryAPI.OperatorDiscoveryResult>() {
```

Comment:
I do not find it too fishy, though. Retrofit does basically the same thing behind the scenes in getAccessTokenFromCode, which is an async retrofit call: retrofit creates a parallel thread and gives you the requested object via the callback... well, after a while. 

I had reasons to want a synchronised retrofit call. This is indicated in retrofit by specifying a returning value rather than a callback parameter, but the thing is that one gets ```android.os.NetworkOnMainThreadException``` if this is to be executed in the main thread. It looks like synchronised retrofit calls are possible to execute only in parallel anyway, which is what I am doing here.